crates/biome_js_analyze/src/lint/nursery/use_import_extensions.rs

```diff
@@ -0,0 +1,222 @@
+    type Query = Ast<AnyJsImportSpecifierLike>;
+    type State = UseImportExtensionsState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+
+        let file_ext = ctx.file_path().extension()?.to_str()?;
+
+        get_extensionless_import(file_ext, node)
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                state.module_name_token.text_range(),
+                markup! {
+                    "Add a file extension for relative imports."
+                },
+            )
+            .note(markup! {
+                "Explicit import improves compatibility with browsers and makes file resolution in tooling faster."
+            }),
+        )
+    }
+
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<JsRuleAction> {
+        let mut mutation = ctx.root().begin();
+
+        let (suggested_path, extension) = state.suggestion.clone()?;
+        let new_module_name = if ctx.as_preferred_quote().is_double() {
+            make::js_string_literal(&suggested_path)
+        } else {
+            make::js_string_literal_single_quotes(&suggested_path)
+        };
+
+        mutation.replace_element(
+            state.module_name_token.clone().into(),
+            new_module_name.into(),
+        );
+
+        Some(JsRuleAction {
+            category: ActionCategory::QuickFix,
+            applicability: Applicability::MaybeIncorrect,
+            message: markup! {
+                "Add potential import extension "<Emphasis>"."{extension}</Emphasis>"."
+            }
+            .to_owned(),
+            mutation,
+        })
+    }
+}
+
+pub struct UseImportExtensionsState {
+    suggestion: Option<(String, String)>,
+    module_name_token: SyntaxToken<JsLanguage>,
+}
+
+fn get_extensionless_import(
+    file_ext: &str,
+    node: &AnyJsImportSpecifierLike,
+) -> Option<UseImportExtensionsState> {
+    let module_name_token = node.module_name_token()?;
+    let module_path = inner_string_text(&module_name_token);
+    let path = Path::new(module_path.text());
+
+    if !matches!(
+        path.components().next(),
+        Some(Component::CurDir | Component::ParentDir)
+    ) || path.extension().is_some()
+    {
+        return None;
+    }
+
+    let has_query_or_hash = path
+        .components()
+        .last()
+        .and_then(|c| c.as_os_str().to_str())
+        .map_or(false, |last| last.contains('?') || last.contains('#'));
+
+    if has_query_or_hash {
+        return Some(UseImportExtensionsState {
+            module_name_token,
+            suggestion: None,
+        });
+    }
+
+    let import_ext = resolve_import_extension(file_ext, path);
+    let mut path_buf = path.to_path_buf();
+
+    let is_index_file = match path.components().last() {
```

Comment:
We could reuse `last_component`
```suggestion
    let is_index_file = match last_component {
```